File: QuickSort.c
#include <stdio.h>

int partition(int numbers[], int lo, int hi) {
  int pivot = numbers[hi];
  int idx = lo - 1;

  for (int i = lo; i < hi; i++) {
    if (numbers[i] <= pivot) {
      idx++;
      int tmp = numbers[idx];
      numbers[idx] = numbers[i];
      numbers[i] = tmp;
    }
  }

  idx++;
  numbers[hi] = numbers[idx];
  numbers[idx] = pivot;

  return idx;
}

void qs(int numbers[], int lo, int hi) {
  if (lo >= hi) {
    return;
  }

  int pivot = partition(numbers, lo, hi);

  qs(numbers, lo, pivot-1);
  qs(numbers, pivot+1, hi);
}

int main() {
  int numbers[] = {2, 8, 5, 9, 2, 7};
  int length = sizeof(numbers) / sizeof(numbers[0]);

  qs(numbers, 0, length-1);

  for (int i = 0; i < length; i++) {
    printf("%d,", numbers[i]);
  }
}

File: quicksort/QuickSort.go
package main

import "fmt"

func partition(numbers []int, lo int, hi int) int {
	pivot := numbers[hi]
	idx := lo - 1

	for i := lo; i < hi; i++ {
		if numbers[i] <= pivot {
			idx++
			tmp := numbers[i]
			numbers[i] = numbers[idx]
			numbers[idx] = tmp
		}
	}

	idx++
	numbers[hi] = numbers[idx]
	numbers[idx] = pivot

	return idx
}

func qs(numbers []int, lo int, hi int) {
	if lo >= hi {
		return
	}

	pivot := partition(numbers, lo, hi)

	qs(numbers, lo, pivot-1)
	qs(numbers, pivot+1, hi)
}

func main() {
	numbers := []int{2, 1, 4, 3, 5}
	qs(numbers, 0, len(numbers)-1)
	fmt.Println(numbers)
}
